ush/get_obs.py

```diff
@@ -0,0 +1,1004 @@
+        set to 6.  This must be between 1 and 24 and must divide evenly into 24.
+
+        hod:
+        The hour of the day.  An integer.  This must be between 0 and 23.  For
+        cumulative fields (CCPA and NOHRSC), hour 0 is treated as that of the
+        next day, i.e. as the 24th hour of the current day.
+
+    Returns:
+        arcv_hr:
+        The hour since the start of day corresponding to the archive file containing
+        the obs file for the given hour of day.  An integer.
+    """
+
+    valid_obtypes = ['CCPA', 'NOHRSC', 'MRMS', 'NDAS']
+    obtype_upper = obtype.upper()
+    if obtype_upper not in valid_obtypes:
+        msg = dedent(f"""
+            The specified observation type (after converting to upper case) is not
+            supported:
+                obtype_upper = {obtype_upper}
+            Valid observation types are:
+                {valid_obtypes}
+        """)
+        logging.error(msg)
+        raise Exception(msg)
+
+    # Ensure that the archive inerval divides evenly into 24 hours.
+    remainder = 24 % arcv_intvl_hrs
+    if remainder != 0:
+        msg = dedent(f"""
+            The archive interval for obs of type {obtype} must divide evenly into 24
+            but doesn't:
+                arcv_intvl_hrs = {arcv_intvl_hrs}
+                24 % arcv_intvl_hrs = {remainder}
+            """)
+        logging.error(msg)
+        raise Exception(msg)
+
+    if (hod < 0) or (hod > 23):
+        msg = dedent(f"""
+            The specified hour-of-day must be between 0 and 23, inclusive, but isn't:
+                hod = {hod}
+        """)
+        logging.error(msg)
+        raise Exception(msg)
+
+    # Set the archive hour.  This depends on the obs type because each obs
+    # type can organize its observation files into archives in a different
+    # way, e.g. a cumulative obs type may put the obs files for hours 1
+    # through 6 of the day in the archive labeled with hour 6 while an
+    # instantaneous obs type may put the obs files for hours 0 through 5 of
+    # the day in the archive labeled with hour 6.
+    if obtype_upper in ['CCPA']:
+        if hod == 0:
+            arcv_hr = 24
+        else:
+            arcv_hr = ceil(hod/arcv_intvl_hrs)*arcv_intvl_hrs
+    elif obtype_upper in ['NOHRSC']:
+        if hod == 0:
+            arcv_hr = 24
+        else:
+            arcv_hr = floor(hod/arcv_intvl_hrs)*arcv_intvl_hrs
+    elif obtype_upper in ['MRMS']:
+        arcv_hr = (floor(hod/arcv_intvl_hrs))*arcv_intvl_hrs
+    elif obtype_upper in ['NDAS']:
+        arcv_hr = (floor(hod/arcv_intvl_hrs) + 1)*arcv_intvl_hrs
+
+    return arcv_hr
+
+
+def get_obs(config, obtype, yyyymmdd_task):
+    """
+    This script checks for the existence of obs files of the specified type
+    at the locations specified by variables in the SRW App's configuration
+    file.  If one or more of these files do not exist, it retrieves them from
```

Comment:
Added a note about this.

One modification that would help reduce hard-coding is that `retrieve_data.py` return the names of the files that it retrieved, so that this script can use them as the raw file names when moving or copying the raw files to their processed locations.  Currently, the raw file names are hard-coded in this script.  I mention this in the comments below.

We can make this change in another PR, though.

Reply:
I agree that would be a great change. The script *does* write out a log file that contains that info but it's hard to parse, it would be great functionality for use in python especially.